internal/topology/variables/cluster_variable_defaulting.go

```diff
@@ -30,66 +30,89 @@ import (
 // If they do not exist yet, they are created if createVariables is set.
-func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassVariable, createVariables bool, fldPath *field.Path) ([]clusterv1.ClusterVariable, field.ErrorList) {
+func defaultClusterVariables(clusterVariables []clusterv1.ClusterVariable, clusterClassVariables []clusterv1.ClusterClassStatusVariable, createVariables bool, fldPath *field.Path) ([]clusterv1.ClusterVariable, field.ErrorList) {
 	var allErrs field.ErrorList
 
-	// Build maps for easier and faster access.
-	clusterVariablesMap := getClusterVariablesMap(clusterVariables)
-	clusterClassVariablesMap := getClusterClassVariablesMap(clusterClassVariables)
-
-	// Validate that all variables in the Cluster are defined in the ClusterClass.
-	// Note: If we don't validate this, we would get a nil pointer dereference below.
-	allErrs = append(allErrs, validateClusterVariablesDefined(clusterVariables, clusterClassVariablesMap, fldPath)...)
-	if len(allErrs) > 0 {
-		return nil, allErrs
+	// Get a map of ClusterVariables and ensure that variables are not defined more than once in Cluster spec.
+	clusterVariablesMap, err := getClusterVariablesMap(clusterVariables)
+	if err != nil {
+		return nil, append(allErrs, field.Invalid(fldPath, clusterVariables,
+			fmt.Sprintf("cluster variables not valid: %s", err)))
 	}
 
+	// Get a map of ClusterClassVariables for each variable name and definition.
+	clusterClassVariablesMap := getClusterClassVariablesMap(clusterClassVariables)
+
 	// allVariables is used to get a full correctly ordered list of variables.
-	allVariables := []string{}
+	allVariables := []clusterv1.ClusterVariable{}
 	// Add any ClusterVariables that already exist.
-	for _, variable := range clusterVariables {
-		allVariables = append(allVariables, variable.Name)
-	}
+	allVariables = append(allVariables, clusterVariables...)
+
 	// Add variables from the ClusterClass, which currently don't exist on the Cluster.
 	for _, variable := range clusterClassVariables {
-		// Continue if the ClusterClass variable already exists.
-		if _, ok := clusterVariablesMap[variable.Name]; ok {
-			continue
+		for _, definition := range variable.Definitions {
+			if _, ok := clusterVariablesMap[variable.Name]; ok {
+				// Continue if the Cluster variable with a definitionFrom this ClusterClass variable exists.
+				if _, ok := clusterVariablesMap[variable.Name][definition.From]; ok {
+					continue
+				}
+				// Continue if the Cluster variable with an empty definitionFrom exists. The user intention here is to
+				// use the default value from a ClusterClass variable with no conflicting variables.
+				if _, ok := clusterVariablesMap[variable.Name][emptyVariableDefinitionFrom]; ok {
+					continue
+				}
+			}
+			allVariables = append(allVariables, clusterv1.ClusterVariable{
+				Name:           variable.Name,
+				DefinitionFrom: definition.From,
```

Comment:
Q: Doesn't this mean that we never add a global variable here? (as definition.From will always have some value)

Reply:
Also seems that when a variable with DefinitionsConflicts: false has to be entirely defauled we will default a value for each DefinitionFrom?

(these sentences would be really easier to read if I could write namespace :joy:)